### PR TITLE
Update Halibut and move AutofacServiceFactory to Tentacle

### PR DESCRIPTION
**Pull request by andrew-at-octopus**
# Background

Halibut has been updated to make the protocol more secure.
In the process, the AutofacServiceFactory has been moved from shared to here.

# Review

Firstly, thanks for reviewing this pull request! :tada:

# Checks

- [ ] :green_heart: All automated builds and tests are passing
- [ ] Scripts that automate Tentacle installation and configuration will continue working after updating to this version of Tentacle
- [ ] Existing installations will continue working after updating to this version of Tentacle
- [ ] I have considered the changes to public documentation needed to reflect this change
- [ ] I have considered the changes to the project wiki needed to reflect this change

### CODE REVIEW

**Comment by andrew-at-octopus** [comment]
[Remote Code Execution via Deserialisation in Halibut Protocol](https://trello.com/c/EZa6ulAd/507-remote-code-execution-via-deserialisation-in-halibut-protocol)

File: linux-packages/packaging-scripts/create-linux-packages.sh [hunk 1]
```diff
@@ -129,14 +129,18 @@ fpm --version "$VERSION" \
...
   "${FPM_RPM_OPTS[@]}" \
   "${FPM_OPTS[@]}" \
   "$INPUT_PATH/=$INSTALL_PATH/" \
   tmp_usr_bin/=/usr/bin/
 set +ex
 
+echo "Signing .rpm package."
+set -ex
+rpmsign --addsign *.rpm
+set +ex
```
Comment (MissedTheMark) [inline linux-packages/packaging-scripts/create-linux-packages.sh:142]: Interesting that the other way stopped working

File: source/Octopus.Tentacle/Services/ServicesModule.cs [hunk 2]
```diff
@@ -15,11 +17,10 @@ protected override void Load(ContainerBuilder builder)
...
 
-            // Find anything with the [Service] attribute
-            builder.RegisterAssemblyTypes(ThisAssembly)
-                .Where(t => t.GetCustomAttributes(typeof (ServiceAttribute), true).Length > 0)
-                .Named(t => "I" + t.Name, typeof (object))
-                .SingleInstance();
+            // Register our Halibut services
+            var serviceTypes = ThisAssembly.GetTypes().Where(t => t.GetCustomAttributes(typeof(ServiceAttribute), true).Length > 0).ToArray();
+            var assemblyServices = new KnownServiceSource(serviceTypes);
+            builder.RegisterInstance(assemblyServices).AsImplementedInterfaces();
```
Comment (MissedTheMark) [inline source/Octopus.Tentacle/Services/ServicesModule.cs:23]: These don't need to be single instances anymore?

**Review by MissedTheMark — COMMENTED** [review]
LGTM generally, would just like confirmation that the registered services don't need to be single instances any more.

File: linux-packages/packaging-scripts/create-linux-packages.sh (hunk 1 above)
Comment (andrew-at-octopus) [reply linux-packages/packaging-scripts/create-linux-packages.sh:142]: Yeah, Debian version was updated, and rpm no longer supports the --sign option

File: source/Octopus.Tentacle/Services/ServicesModule.cs (hunk 2 above)
Comment (andrew-at-octopus) [reply source/Octopus.Tentacle/Services/ServicesModule.cs:23]: Pretty sure RegisterInstance takes care of that :)

Comment (MissedTheMark) [reply source/Octopus.Tentacle/Services/ServicesModule.cs:23]: lol yes 😅

**Review by MissedTheMark — APPROVED** [review]
LGTM 👍

**Pull request merged**